### ISSUE TITLE
Broken development version 2.5.9d1 accidentally released to WP Plugin Directory

### ISSUE THREAD

**Issue by Rumperuu** [markcheret/footnotes#55]
This Issue serves as a record of the accidental release a broken development version of the Plugin to the WP Plugin Directory on 2021-03-01, which caused many users' sites to break. This was a serious error, necessitating an immediate post-mortem and detailing of steps to be taken to avoid it happening again.

A full post-mortem will be posted on the [WP Support Forum](https://wordpress.org/support/plugin/footnotes/) at 22:00 (UK time).

### Bug Reports

- https://wordpress.org/support/plugin/footnotes/
- https://wordpress.org/support/topic/latest-update-broke-my-site-19/
- https://wordpress.org/support/topic/version-2-5-9d1-breaks-wp-down/
- https://wordpress.org/support/topic/crashed-my-site-104/
- https://wordpress.org/support/topic/update-breaks-layout-3/
- https://wordpress.org/support/topic/bug-history/
- direct emails to @markcheret 

### Initial Response

- [Pinned post](https://wordpress.org/support/topic/2-5-10-reverts-2-5-9d1-and-apologies/) by @markcheret

**Comment by Rumperuu** [comment]
As this was a serious issue, I have conducted a full post-mortem. This issue has also been tracked on the GitHub repo. <a href="https://github.com/markcheret/footnotes/issues/55" rel="noopener" target="_blank">here</a>.

<strong>Issue</strong>

As part of <a href="https://github.com/markcheret/footnotes/pull/50" rel="noopener" target="_blank">refactoring the codebase</a> to comply with the <a href="https://codex.wordpress.org/WordPress_Coding_Standards" rel="noopener" target="_blank">WordPress Coding Standards</a> (the first step in <a href="https://github.com/markcheret/footnotes/projects/3" rel="noopener" target="_blank">a wider project</a> of migration and workflow automation), the following file and directory renames were made:

<ul>
<li>`class/dashboard/` → `class/layout/`;</li>
<li>`class/dashboard/layout.php` → `class/layout/abstract-engine.php`; </li>
<li>`class/dashboard/subpage-diagnostics.php` → `class/layout/diagnostics.php`; and</li>
<li>`class/dashboard/subpage-main.php` → `class/layout/settings.php`.</li>
</ul>

To account for these relocations, the penultimate line of `includes.php` was changed from `mci_footnotes_require_php_files( dirname( __FILE__ ) . '/class/dashboard' );` to `mci_footnotes_require_php_files( dirname( __FILE__ ) . '/class/layout' );` to ensure that the new files were found. 

In the v2.5.9d1 release, the updated `includes.php` was included (see <a href="https://plugins.svn.wordpress.org/footnotes/tags/2.5.9d1/includes.php" rel="noopener" target="_blank">here</a>), but the `class/dashboard/` → `class/layout/` directory rename was not (see <a href="https://plugins.svn.wordpress.org/footnotes/tags/2.5.9d1/class/dashboard/" rel="noopener" target="_blank">here</a>).

<strong>History (all times given are UK time)</strong>

`trunk/` was originally on r2483011 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=r2483011" rel="noopener" target="_blank">here</a>), with a ‘Stable Tag’ of 2.5.8 (see `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483011" rel="noopener" target="_blank">here</a>).

At 07:25 today I pushed a branch called `pre_standardisation` from the code in `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483011" rel="noopener" target="_blank">here</a>).
 
At 07:37 I committed revision r2483336 to `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=2483336" rel="noopener" target="_blank">here</a>). This contained the entire newly-updated codebase, including the normalised line endings which made the file diff unusable. I also updated the 2 version numbers in `footnotes.php` to 2.5.9d1 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/footnotes.php?p=2483336" rel="noopener" target="_blank">here</a>) along with the ‘Stable Tag’ in `readme.txt` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483336" rel="noopener" target="_blank">here</a>). This commit correctly updated `includes.php` to search for the new `layout/` folder (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/includes.php?p=2483336" rel="noopener" target="_blank">here</a>), but did not correctly update `class/dashboard/` -> `class/layout/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/class/dashboard/?p=2483336" rel="noopener" target="_blank">here</a>).

At 07:54 I reverted all of the previous commit except for the line ending changes (r2483350, see here). For an unknown reason, this reverted the ‘Stable Tag’ field to 2.5.5 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483350" rel="noopener" target="_blank">here</a>; it's possible that it reverted 4 commits back to r2478015 instead (compare that commit's `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2478015" rel="noopener" target="_blank">here</a>) but I can see no reason why this should have happened).

At 07:57 I committed revision r2483354 to `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=r2483354" rel="noopener" target="_blank">here</a>), containing all of the standardisation tweaks minus the line ending changes and again setting the three version numbers to 2.5.9d1 (see `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2483354" rel="noopener" target="_blank">here</a>).

The above steps are recorded in my `.bash_history` as follows:

```
1944 svn co https://plugins.svn.wordpress.org/footnotes/ footnotes-svn                                   
1945 cd footnotes-svn/                                                              
1946 svn copy trunk/ branches/pre_standardisation                                   
1947 svn status                                                                     
1948 svn commit -m "Creating branch of pre-standardisation codebase in case anything goes wrong"
1949 ls                                                                             
1950 cd ..                                                                          
1951 cd footnotes\ \(svn\)/ 
```
[Here I deleted all the files in my local copy of `trunk/` in a GUI File Viewer and copied over the files from my local checkout of the Git repo's `main` branch @ commit <a href="https://github.com/markcheret/footnotes/commit/2e0b2f8d98565e5cd990f31d50023c9c6f9a5e3f" rel="noopener" target="_blank">2e0b2f8</a>]
[I then copied the files in my local copy of `trunk/` into my local WordPress development server's Plugins folder to test, and everything worked fine]       
```                                  
1952 svn stat                                                                       
1953 svn diff trunk/footnotes.php                                                   
1954 svn ci -m "Standardise codebase to WP Coding Standards"  
```
[Here I did that again, this time copying over the files from my local copy of `main` at commit <a href="https://github.com/markcheret/footnotes/commit/8d0ebdb8595ab698c50996bc268bcabae928294d" rel="noopener" target="_blank">8d0ebdb</a> (i.e., the commit that normalised the line endings)] 
```                         
1955 svn stat                                                                       
1956 svn ci -m "Revert standardise codebase to WP Coding Standards"
```
[Here I replaced the contents of my local `trunk/` a third time, this time copying over the files from `main` @ commit <a href="https://github.com/markcheret/footnotes/commit/2e0b2f8d98565e5cd990f31d50023c9c6f9a5e3f" rel="noopener" target="_blank">2e0b2f8</a> again]
```
1957 svn ci -m "Re-standardise codebase to WP Coding Standards (this time without the line ending changes)"
```

I followed the guide to updating the repo. from the WP Plugin Handbook (see <a href="https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#editing-existing-files" rel="noopener" target="_blank">here</a>).

I did nothing further on the SVN repo. At 08:20 I emailed the rest of the project team to say I had updated `trunk/`.

At 11:09 I received an email from the Plugin Directory informing me that @pewgeuges had committed a new 2.5.10 release.

At 15:23 I received an email from my own WordPress site informing me that my copy of footnotes had automatically upgraded from 2.5.8 to 2.5.10.

At 15:33 I ran `svn up` to pull the latest copy of the SVN repo, which included new tagged releases for both 2.5.9d1 and 2.5.10.

<strong>What Happened</strong>

<ol>
<li>I did not realise that the steps given in the WP Plugin Handbook section linked above were only for ‘Editing Existing Files’. `class/layout/` was present in my local copy of `trunk/`, but as the new `class/layout/` directory was not yet added to tracking (i.e., through `svn add`), it was not included in the `svn stat` or the commit, whilst the updated version of `includes.php` (which pointed to it instead of `class/dashboard/`) was;</li>
<li>I did not notice this in my test environment because copying the files from my local copy of `trunk/` meant that I included my local (but un-committed) copy of `class/layout/`;</li>
<li>this would have been noticed as soon as anyone else pulled the remote `trunk/` to their local copy and tested it; but</li> 
<li>the code in `trunk/` was prematurely pushed into a 2.5.9d1 release (i.e., copied to `tags/2.5.9d1`):
<ul>
<li>if this was manually done by someone, then it was not tested locally beforehand;</li>
<li>however, as we'd previously discussed the fact that the codebase standardisation would not merit a new release (see <a href="https://github.com/markcheret/footnotes/pull/50#issuecomment-787485248" rel="noopener" target="_blank">here</a>), I assume this was not the case;</li>
<li>in which case, the WP Plugin Directory may have noticed that the ‘Stable Tag’ field in `readme.txt` was updated and automatically pushed a new release;</li>
<li>this appears to be the case according to the Plugin Handbook (see <a href="https://github.com/markcheret/footnotes/pull/50#issuecomment-787485248" rel="noopener" target="_blank">here</a>), which states that the Directory parses the ‘Stable Tag’ field in `trunk/readme.txt` and, if the value is set to something other than ‘trunk’, will look in `tags/<stable_tag>`;</li>
<li>per the Handbook, ‘If there is no folder for the Stable Tag, then WordPress will default to using the trunk folder’s content’;</li>
<li>so the Directory must have parsed `trunk/readme.txt`, read in the ‘Stable Tag’ of 2.5.9d1, looked for `tags/2.5.9d1`, not found it and then pushed its own 2.5.9d1 release rather than falling back to `trunk/`.</li>
</ul></li>
</ol>

<strong>What Went Wrong</strong>

<ol>
<li>I was unaware that the WP Plugin Directory parsed `trunk/readme.txt` rather than looking at the latest version release in `tags/` — I thought that it was safe to put pre-release code in `trunk/`, and that the version numbers (two in `footnotes.php` and one in `readme.txt`) should be updated to reflect the current version (in this case 2.5.9d1), whereas in fact the ‘Stable Tag’ should remain the latest release version (see example 2.5.6d4 release with 2.5.5 version number <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2478015" rel="noopener" target="_blank">here</a>);</li>
<li>unfamiliarity with SVN led to me failing to commit new, untracked files (i.e., `class/layout/`) whilst correctly committing the edited tracked files that referred to them (i.e., `includes.php`);</li>
<li>my own manual testing failed to detect this due to copying local files into the test environment rather than pulling files from the remote repo; but</li>
<li>as the Directory pushed a new release automatically, there was no time for anybody else to test the changes and notice the error; and</li>
<li>as the 2.5.9d1 release was quickly followed up with 2.5.10 by @Marcel, I did not receive an email notification about my own site upgrading to footnotes 2.5.9d1 and realise that there was a problem earlier; and, finally</li>
<li>as I am unavailable this week due to work commitments I have set up an inbox filter automatically moving forum post emails out of inbox, so I was not aware of the flood of forum bug reports.</li>
</ol>
 
<strong>Impact Assessment</strong>

This was entirely my mistake due a misunderstanding of how the SVN repo. behaved. It was swiftly fixed by @pewgeuges, but had a significant impact on users (as evidenced by the number of Support Forum posts). The damage appears to be limited to users who have enabled automatic Plugin updates, although as ~50% of the userbase are using the most recent Plugin version, we can infer that potentially as many as 50% of users have enabled this (though the real figure is likely to be smaller). Thankfully, many of those same users will have automatically upgraded to 2.5.10, but some may have disabled automatic updates whilst the Plugin was broken. Hopefully, they will soon see @markcheret's message and upgrade.

<strong>Mitigations</strong>

<ol>
<li>Do not update the ‘Stable Tag’ field in `trunk/readme.txt` for development versions to ensure that committing broken code to `trunk/` will not accidentally trigger a release;</li>
<li>add release instructions for future contributors (see <a href="https://github.com/markcheret/footnotes/wiki/SVN-Release-Guide" rel="noopener" target="_blank">here</a>) to ensure that the same mistakes are not made by others;</li>
<li>set up an automated release workflow (see <a href="https://github.com/markcheret/footnotes/issues/30" rel="noopener" target="_blank">here</a>) in order to completely remove the risk of human error.</li>
<ol>

Thank you all for readying, and my sincere apologies to anyone inconvenienced by this issue,
Ben

**Comment by Rumperuu** [comment]
<strong>What Happened</strong>

<ol>
<li>I did not realise that the steps given in the WP Plugin Handbook section linked above were only for ‘Editing Existing Files’. `class/layout/` was present in my local copy of `trunk/`, but as the new `class/layout/` directory was not yet added to tracking (i.e., through `svn add`), it was not included in the `svn stat` or the commit, whilst the updated version of `includes.php` (which pointed to it instead of `class/dashboard/`) was;</li>
<li>I did not notice this in my test environment because copying the files from my local copy of `trunk/` meant that I included my local (but un-committed) copy of `class/layout/`;</li>
<li>this would have been noticed as soon as anyone else pulled the remote `trunk/` to their local copy and tested it; but</li> 
<li>the code in `trunk/` was prematurely pushed into a 2.5.9d1 release (i.e., copied to `tags/2.5.9d1`):
<ul>
<li>if this was manually done by someone, then it was not tested locally beforehand;</li>
<li>however, as we'd previously discussed the fact that the codebase standardisation would not merit a new release (see <a href="https://github.com/markcheret/footnotes/pull/50#issuecomment-787485248" rel="noopener" target="_blank">here</a>), I assume this was not the case;</li>
<li>in which case, the WP Plugin Directory may have noticed that the ‘Stable Tag’ field in `readme.txt` was updated and automatically pushed a new release;</li>
<li>this appears to be the case according to the Plugin Handbook (see <a href="https://github.com/markcheret/footnotes/pull/50#issuecomment-787485248" rel="noopener" target="_blank">here</a>), which states that the Directory parses the ‘Stable Tag’ field in `trunk/readme.txt` and, if the value is set to something other than ‘trunk’, will look in `tags/<stable_tag>`;</li>
<li>per the Handbook, ‘If there is no folder for the Stable Tag, then WordPress will default to using the trunk folder’s content’;</li>
<li>so the Directory must have parsed `trunk/readme.txt`, read in the ‘Stable Tag’ of 2.5.9d1, looked for `tags/2.5.9d1`, not found it and then pushed its own 2.5.9d1 release rather than falling back to `trunk/`.</li>
</ul></li>
</ol>

<strong>What Went Wrong</strong>

<ol>
<li>I was unaware that the WP Plugin Directory parsed `trunk/readme.txt` rather than looking at the latest version release in `tags/` — I thought that it was safe to put pre-release code in `trunk/`, and that the version numbers (two in `footnotes.php` and one in `readme.txt`) should be updated to reflect the current version (in this case 2.5.9d1), whereas in fact the ‘Stable Tag’ should remain the latest release version (see example 2.5.6d4 release with 2.5.5 version number <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2478015" rel="noopener" target="_blank">here</a>);</li>
<li>unfamiliarity with SVN led to me failing to commit new, untracked files (i.e., `class/layout/`) whilst correctly committing the edited tracked files that referred to them (i.e., `includes.php`);</li>
<li>my own manual testing failed to detect this due to copying local files into the test environment rather than pulling files from the remote repo; but</li>
<li>as the Directory pushed a new release automatically, there was no time for anybody else to test the changes and notice the error; and</li>
<li>as the 2.5.9d1 release was quickly followed up with 2.5.10 by @Marcel, I did not receive an email notification about my own site upgrading to footnotes 2.5.9d1 and realise that there was a problem earlier; and, finally</li>
<li>as I am unavailable this week due to work commitments I have set up an inbox filter automatically moving forum post emails out of inbox, so I was not aware of the flood of forum bug reports.</li>
</ol>
 
<strong>Impact Assessment</strong>

This was entirely my mistake due a misunderstanding of how the SVN repo. behaved. It was swiftly fixed by @pewgeuges, but had a significant impact on users (as evidenced by the number of Support Forum posts). The damage appears to be limited to users who have enabled automatic Plugin updates, although as ~50% of the userbase are using the most recent Plugin version, we can infer that potentially as many as 50% of users have enabled this (though the real figure is likely to be smaller). Thankfully, many of those same users will have automatically upgraded to 2.5.10, but some may have disabled automatic updates whilst the Plugin was broken. Hopefully, they will soon see @markcheret's message and upgrade.

<strong>Mitigations</strong>

<ol>
<li>Do not update the ‘Stable Tag’ field in `trunk/readme.txt` for development versions to ensure that committing broken code to `trunk/` will not accidentally trigger a release;</li>
<li>add release instructions for future contributors (see <a href="https://github.com/markcheret/footnotes/wiki/SVN-Release-Guide" rel="noopener" target="_blank">here</a>) to ensure that the same mistakes are not made by others;</li>
<li>set up an automated release workflow (see <a href="https://github.com/markcheret/footnotes/issues/30" rel="noopener" target="_blank">here</a>) in order to completely remove the risk of human error.</li>
<ol>

Thank you all for readying, and my sincere apologies to anyone inconvenienced by this issue,
Ben

**Comment by Rumperuu** [comment]
As this was a serious issue, I have conducted a full post-mortem. A general-purpose summary has been shared on the [WP Support Forum](https://wordpress.org/support/topic/2-5-10-reverts-2-5-9d1-and-apologies/#post-14119440).

### Issue

As part of <a href="https://github.com/markcheret/footnotes/pull/50" rel="noopener" target="_blank">refactoring the codebase</a> to comply with the <a href="https://codex.wordpress.org/WordPress_Coding_Standards" rel="noopener" target="_blank">WordPress Coding Standards</a> (the first step in <a href="https://github.com/markcheret/footnotes/projects/3" rel="noopener" target="_blank">a wider project</a> of migration and workflow automation), the following file and directory renames were made:

- `class/dashboard/` → `class/layout/`;
- `class/dashboard/layout.php` → `class/layout/abstract-engine.php`; 
- `class/dashboard/subpage-diagnostics.php` → `class/layout/diagnostics.php`; and
- `class/dashboard/subpage-main.php` → `class/layout/settings.php`.

To account for these relocations, the penultimate line of `includes.php` was changed from `mci_footnotes_require_php_files( dirname( __FILE__ ) . '/class/dashboard' );` to `mci_footnotes_require_php_files( dirname( __FILE__ ) . '/class/layout' );` to ensure that the new files were found. 

In the v2.5.9d1 commit to `trunk/`, the updated `includes.php` was included (see <a href="https://plugins.svn.wordpress.org/footnotes/tags/2.5.9d1/includes.php" rel="noopener" target="_blank">here</a>), but the `class/dashboard/` → `class/layout/` directory rename was not (see <a href="https://plugins.svn.wordpress.org/footnotes/tags/2.5.9d1/class/dashboard/" rel="noopener" target="_blank">here</a>).

`includes.php` tried to find files in a non-existent directory, couldn’t, and threw a fatal error.

### History (all times given are UK time)

`trunk/` was originally on r2483011 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=r2483011" rel="noopener" target="_blank">here</a>), with a ‘Stable Tag’ of 2.5.8 (see `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483011" rel="noopener" target="_blank">here</a>).

At 07:25 today I pushed a branch called `pre_standardisation` from the code in `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483011" rel="noopener" target="_blank">here</a>).
 
At 07:37 I committed revision r2483336 to `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=2483336" rel="noopener" target="_blank">here</a>). This contained the entire newly-updated codebase, including the normalised line endings which made the file diff unusable. I also updated the 2 version numbers in `footnotes.php` to 2.5.9d1 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/footnotes.php?p=2483336" rel="noopener" target="_blank">here</a>) along with the ‘Stable Tag’ in `readme.txt` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483336" rel="noopener" target="_blank">here</a>). This commit correctly updated `includes.php` to search for the new `layout/` folder (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/includes.php?p=2483336" rel="noopener" target="_blank">here</a>), but did not correctly update `class/dashboard/` -> `class/layout/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/class/dashboard/?p=2483336" rel="noopener" target="_blank">here</a>).

At 07:54 I reverted all of the previous commit except for the line ending changes (r2483350, see here). For an unknown reason, this reverted the ‘Stable Tag’ field to 2.5.5 (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=2483350" rel="noopener" target="_blank">here</a>; it's possible that it reverted 4 commits back to r2478015 instead (compare that commit's `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2478015" rel="noopener" target="_blank">here</a>) but I can see no reason why this should have happened).

At 07:57 I committed revision r2483354 to `trunk/` (see <a href="https://plugins.svn.wordpress.org/footnotes/trunk/?p=r2483354" rel="noopener" target="_blank">here</a>), containing all of the standardisation tweaks minus the line ending changes and again setting the three version numbers to 2.5.9d1 (see `readme.txt` <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2483354" rel="noopener" target="_blank">here</a>).

The above steps are recorded in my `.bash_history` as follows:
```
1944 svn co https://plugins.svn.wordpress.org/footnotes/ footnotes-svn                                   
1945 cd footnotes-svn/                                                              
1946 svn copy trunk/ branches/pre_standardisation                                   
1947 svn status                                                                     
1948 svn commit -m "Creating branch of pre-standardisation codebase in case anything goes wrong"
1949 ls                                                                             
1950 cd ..                                                                          
1951 cd footnotes\ \(svn\)/ 
```
Here I deleted all the files in my local copy of `trunk/` in a GUI File Viewer and copied over the files from my local checkout of the Git repo's `main` branch @ commit <a href="https://github.com/markcheret/footnotes/commit/2e0b2f8d98565e5cd990f31d50023c9c6f9a5e3f" rel="noopener" target="_blank">2e0b2f8</a>.

I then copied the files in my local copy of `trunk/` into my local WordPress development server's Plugins folder to test, and everything worked fine.
```                                  
1952 svn stat                                                                       
1953 svn diff trunk/footnotes.php                                                   
1954 svn ci -m "Standardise codebase to WP Coding Standards"  
```
Here I did that again, this time copying over the files from my local copy of `main` at commit <a href="https://github.com/markcheret/footnotes/commit/8d0ebdb8595ab698c50996bc268bcabae928294d" rel="noopener" target="_blank">8d0ebdb</a> (i.e., the commit that normalised the line endings).
```                         
1955 svn stat                                                                       
1956 svn ci -m "Revert standardise codebase to WP Coding Standards"
```
Here I replaced the contents of my local `trunk/` a third time, this time copying over the files from `main` @ commit <a href="https://github.com/markcheret/footnotes/commit/2e0b2f8d98565e5cd990f31d50023c9c6f9a5e3f" rel="noopener" target="_blank">2e0b2f8</a> again.
```
1957 svn ci -m "Re-standardise codebase to WP Coding Standards (this time without the line ending changes)"
```

I followed the guide to updating the repo. from the WP Plugin Handbook (see <a href="https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#editing-existing-files" rel="noopener" target="_blank">here</a>).

I did nothing further on the SVN repo. At 08:20 I emailed the rest of the project team to say I had updated `trunk/`.

At 11:09 I received an email from the Plugin Directory informing me that @pewgeuges had committed a new 2.5.10 release.

At 15:23 I received an email from my own WordPress site informing me that my copy of footnotes had automatically upgraded from 2.5.8 to 2.5.10.

At 15:33 I ran `svn up` to pull the latest copy of the SVN repo, which included new tagged releases for both 2.5.9d1 and 2.5.10.

### What Happened

1. I did not realise that the steps given in the WP Plugin Handbook section linked above were only for ‘Editing Existing Files’. `class/layout/` was present in my local copy of `trunk/`, but as the new `class/layout/` directory was not yet added to tracking (i.e., through `svn add`), it was not included in the `svn stat` or the commit, whilst the updated version of `includes.php` (which pointed to it instead of `class/dashboard/`) was;
1. I did not notice this in my test environment because copying the files from my local copy of `trunk/` meant that I included my local (but un-committed) copy of `class/layout/`;
1. this would have been noticed as soon as anyone else pulled the remote `trunk/` to their local copy and tested it; but 
1. the code in `trunk/` was taken by the WP Plugin Directory to be a new release (2.5.9d1):
    - per the WP Plugin Handbook (see <a href="https://github.com/markcheret/footnotes/pull/50#issuecomment-787485248" rel="noopener" target="_blank">here</a>), the Directory parses the ‘Stable Tag’ field in `trunk/readme.txt` first and, if the value is set to something other than ‘trunk’, will look in `tags/⟨stable_tag⟩`;
    - again per the Handbook, ‘If there is no folder for the Stable Tag, then WordPress will default to using the trunk folder’s content’;
    - so the Directory parsed `trunk/readme.txt`, read in the ‘Stable Tag’ of 2.5.9d1, looked for `tags/2.5.9d1`, couldn't find it and so pushed the contents of `trunk/` our as a new release 2.5.9d1.
1. as the codebase standardisation changeset covered 45 files, the commit diff filelist was too long for anyone to notice that the `class/layout/` directory was missing.

### What Went Wrong

1. I was unaware that the WP Plugin Directory parsed `trunk/readme.txt` rather than looking at the latest version release in `tags/` — I thought that it was safe to put pre-release code in `trunk/`, and that the version numbers (two in `footnotes.php` and one in `readme.txt`) should be updated to reflect the current version (in this case 2.5.9d1), whereas in fact the ‘Stable Tag’ should remain the latest release version (see example 2.5.6d4 release with 2.5.5 version number <a href="https://plugins.svn.wordpress.org/footnotes/trunk/readme.txt?p=r2478015" rel="noopener" target="_blank">here</a>);
1. @pewgeuges did not know that I was not aware of this (see [here](https://github.com/markcheret/footnotes/pull/50#issuecomment-787624123));
1. unfamiliarity with SVN led to me failing to commit new, untracked files (i.e., `class/layout/`) whilst correctly committing the edited tracked files that referred to them (i.e., `includes.php`);
1. my own manual testing failed to detect this due to copying local files into the test environment rather than pulling files from the remote repo; but
1. as the Directory pushed a new release automatically, there was no time for anybody else to test the changes and notice the error; and
1. as the 2.5.9d1 release was quickly followed up with 2.5.10 by @pewgeuges, I did not receive an email notification about my own site upgrading to footnotes 2.5.9d1 and realise that there was a problem earlier; and, finally
1. as I am unavailable this week due to work commitments I have set up an inbox filter automatically moving forum post emails out of inbox, so I was not aware of the flood of forum bug reports.
 
### Impact Assessment

This was entirely my mistake due a misunderstanding of how the SVN repo. behaved. It was swiftly fixed by @pewgeuges, but had a significant impact on users (as evidenced by the number of Support Forum posts). The damage appears to be limited to users who have enabled automatic Plugin updates, although as ~50% of the userbase are using the most recent Plugin version, we can infer that potentially as many as 50% of users have enabled this (though the real figure is likely to be smaller). Thankfully, many of those same users will have automatically upgraded to 2.5.10, but some may have disabled automatic updates whilst the Plugin was broken. Hopefully, they will soon see @markcheret's message and upgrade.

### Mitigations

1. Do not update the ‘Stable Tag’ field in `trunk/readme.txt` for development versions to ensure that committing broken code to `trunk/` will not accidentally trigger a release;
1. add release instructions for future contributors (see <a href="https://github.com/markcheret/footnotes/wiki/SVN-Release-Guide" rel="noopener" target="_blank">here</a>) to ensure that the same mistakes are not made by others;
1. ensure that commits are kept small to increase the likelihood of errors being spotted (see [here](https://github.com/markcheret/footnotes/pull/50#issuecomment-787750474));
1. set up an automated release workflow (see <a href="https://github.com/markcheret/footnotes/issues/30" rel="noopener" target="_blank">here</a>) in order to completely remove the risk of human error.

Thank you for reading, and my sincere apologies to anyone inconvenienced by this issue,
Ben
